### PR TITLE
Completed User can save venue to favorites - Mohamed

### PR DESCRIPTION
**Pull request by Mohamed-Elmessiry**
Screen Shot 2021-06-09 at 5 50 39 PM

### CODE REVIEW

File: client/pages/home.jsx [hunk 1]
```diff
@@ -6,39 +6,64 @@ const App = () => {
...
 
       fetch(`/api/venue/image/${venue.id}`)
         .then(data => data.json())
         .then(res => {
           const newImageLinks = imageLinks.concat({ id: venue.id, url: res.imgUrl });
+
           if (res.imgUrl) { venues[i].imageUrl = res.imgUrl; } else {
-            venues[i].imageUrl = 'https://cdn.stocksnap.io/img-thumbs/960w/sliced-homemade_5BADCUBZS9.jpg';
+
+            venues[i].imageUrl = 'http://localhost:3001/images/noimage.jpeg';
```
Comment (thebearingedge) [inline client/pages/home.jsx]: you can just put `/images/noimage.jpeg` instead of the whole URL. Otherwise, this will break when you deploy your app.

File: client/pages/venue.jsx [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import React from 'react';
 
 const Venue = () => {
+  // eslint-disable-next-line no-unused-vars
```
Comment (thebearingedge) [inline client/pages/venue.jsx]: I don't think this comment is necessary.

File: client/pages/venue.jsx [hunk 3]
```diff
@@ -10,21 +12,34 @@ const Venue = () => {
...
+
   const onSave = e => {
-    alert('save');
+    const requestOptions = {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(venue)
+    };
+    fetch('/api/user/addFavorite/', requestOptions).then(res => {
+      alert(res);
```
Comment (thebearingedge) [inline client/pages/venue.jsx]: Please remove all alerts.

File: client/pages/venue.jsx [hunk 4]
```diff
@@ -10,21 +12,34 @@ const Venue = () => {
...
   };
   const onMenu = e => {
     window.open(venue.menuUrl);
   };
   const mainPageClicked = e => {
+    localStorage.removeItem('favorites');
+    window.location.replace('/');
+  };
+  const favorites = e => {
+    localStorage.setItem('favorites', true);
```
Comment (thebearingedge) [inline client/pages/venue.jsx]: If you are using `react-router`, then you should use their `<Link />` component instead of onClick.

File: server/index.js [hunk 5]
```diff
@@ -33,13 +35,14 @@ const formatData = json => {
       };
     });
 };
-
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: true }));
```
Comment (thebearingedge) [inline server/index.js]: What is line 39 for?

File: server/index.js [hunk 6]
```diff
@@ -68,7 +81,52 @@ app.get('/api/venues/:city/:search', (req, res, next) => {
       res.send(formatData(json));
     });
 });
+app.get('/create', (req, res, next) => {
+  const db = new sqlite.Database('./tripster.db');
 
+  db.run("DELETE FROM favorites WHERE venue_json = '456'");
+  db.close();
+  res.end('created');
```
Comment (thebearingedge) [inline server/index.js]: What's wrong with using PostgreSQL? sqlite won't work reliably on Heroku

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: client/pages/venue.jsx (hunk 2 above)
Comment (Mohamed-Elmessiry) [reply client/pages/venue.jsx]: It's necessary, as it underlines the line underneath, which doesn't allow me to commit.

Comment (thebearingedge) [reply client/pages/venue.jsx]: then you can delete the variable you are not using 😄

File: package.json [hunk 7]
```diff
@@ -34,11 +34,13 @@
     "express": "^4.17.1",
     "file-loader": "^6.2.0",
     "jsonwebtoken": "^8.5.1",
+    "mysql2": "^2.2.5",
```
Comment (thebearingedge) [inline package.json:37]: If you are using mysql, then you can remove `pg` and `sqlite3` from your dependencies.

File: server/db.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+const db = new Pool({
+  user: 'postgres',
+  host: 'localhost',
+  database: 'postgres',
+  schema: 'postgres',
+  password: '',
+  port: 5434
+});
+
+module.exports = db;
```
Comment (thebearingedge) [inline server/db.js]: do you still need this?

File: server/create_tables.sql [hunk 9]
```diff
@@ -0,0 +1,3 @@
+favorites(username, venue_json)
+post
+mock
```
Comment (thebearingedge) [inline server/create_tables.sql]: is this file needed?

File: server/index.js [hunk 10]
```diff
@@ -4,12 +4,26 @@ require('dotenv/config');
...
+  password: 'Trip1234',
+  database: 'tripster',
+  port: '8889'
+});
 
+connection.connect(function (err) {
+  if (err) {
+    console.error(err);
+  }
+});
```
Comment (thebearingedge) [inline server/index.js]: Is this 22-26 needed?

File: server/index.js [hunk 11]
```diff
@@ -51,7 +65,17 @@ app.get('/api/venue/image/:id?', (req, res, next) => {
...
 });
+app.get('/api/venue/:id', (req, res, next) => {
+  const id = req.params.id;
+  const uri = `https://api.foursquare.com/v2/venues/${id}/?client_id=${process.env.clientId}&client_secret=${process.env.clientSecret}&v=20210514&limit=4`;
 
+  fetch(uri)
+    .then(data => data.json())
+    .then(json => {
+      res.send(json);
+      res.end();
```
Comment (thebearingedge) [inline server/index.js]: `res.send()` already sends a response, you can remove `res.end()`

File: server/index.js [hunk 12]
```diff
@@ -68,6 +92,95 @@ app.get('/api/venues/:city/:search', (req, res, next) => {
       res.send(formatData(json));
     });
 });
+app.get('/create', (req, res, next) => {
+
+  connection.query("DELETE FROM favorites WHERE venue_json = '456'", function (err) {
+    if (err) {
+      console.error(err);
+    }
```
Comment (thebearingedge) [inline server/index.js]: If there is an error, you should still send a 500 response, just like in all of our express exercises. check all of your server code to make sure that the client receives a 500 response when there is an error like this

File: server/index.js [hunk 13]
```diff
@@ -68,6 +92,95 @@ app.get('/api/venues/:city/:search', (req, res, next) => {
...
+
+app.get('/api/get/allposts', (req, res, next) => {
+  connection.query('SELECT * FROM posts ORDER BY date_created DESC', (qErr, qRes) => {
+    if (qRes && qRes.rows) {
+      res.json(qRes.rows);
+    }
+
+    if (qErr) {
+      console.error(qErr);
+      res.json({ err: qErr });
```
Comment (thebearingedge) [inline server/index.js]: never send server errors to the client it's a security risk, only a 500 with a simple message

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: server/index.js (hunk 10 above)
Comment (Mohamed-Elmessiry) [reply server/index.js]: It's needed for mySQL

Comment (thebearingedge) [reply server/index.js]: Does it work if you delete this?

File: server/index.js [hunk 14]
```diff
@@ -60,14 +76,118 @@ app.get('/api/venues/:city/:search', (req, res, next) => {
...
+
+  const str = '{}';
+  const params = ['tester', str];
+  connection.query('INSERT INTO favorites (username, venue_json) VALUES (?,?)', params,
+    (qErr, qRes) => {
+      if (qErr) {
+        console.error(qErr);
+        res.status(404).send('requested info not found');
+      }
+      res.json(qRes.rows);
```
Comment (thebearingedge) [inline server/index.js]: be sure to include a `return` statement or `else` so that it is impossible for your server to try to respond twice to the same request. check all of your server code for this.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

**Review by thebearingedge — APPROVED** [review]
